src/pydvl/parallel/backends/joblib.py

```diff
@@ -25,14 +25,29 @@ class JoblibParallelBackend(BaseParallelBackend, backend_name="joblib"):
     It shouldn't be initialized directly. You should instead call
     [init_parallel_backend()][pydvl.parallel.backend.init_parallel_backend].
 
+    ??? Example
+        ``` python
+        from pydvl.parallel import init_paralle_backend, ParallelConfig
+        config = ParallelConfig(backend="joblib")
+        parallel_backend = init_parallel_backend(config)
+        ```
+
+    ??? Example
+        ``` python
+        import joblib
+        from pydvl.parallel import init_paralle_backend, ParallelConfig
+        with joblib.parallel_config(verbose=100):
+            config = ParallelConfig(backend="joblib")
+            parallel_backend = init_parallel_backend(config)
```

Comment:
The suggested pattern is very similar to [sklearn's parallelization pattern](https://scikit-learn.org/stable/computing/parallelism.html), except for the parallel backend thing, so I would think that users would be familiar with it.

I preferred this pattern over hiding joblib completely away in our class because we would have to duplicate its entire configuration interface and keep passing the values around. I didn't like it from a maintainability point of view.

You're right about the private constructors. I don't remember why we needed them as well. I was going to remove them in a follow-up PR anyway.

Reply:
I guess if the ParallelConfig line goes away, it's only two lines, which is better. But still, having our init and their init is very weird. Can we do our init from theirs?
```python
with joblib.parallel_config(blah) as pc:
    pb = JoblibParallelBackend(pc)
    compute_schtuff(u, params, parallel_backend=pb)
```
Because the discussion about just using joblib as our backend and using it to talk to dask/ray/whatever ended up showing that we couldn't, correct?

Reply:
With the change I want to do in the follow-up PR this is how it would look like:

```python
with joblib.parallel_config(blah):
    pb = JoblibParallelBackend()
    compute_schtuff(u, params, parallel_backend=pb)
```

We only use `joblib` inside `MapReduceJob` and the change in this PR makes it so that the users don't have to pass joblib parameters to our classes and instead rely on joblib's mechanisms to do that.

The idea of just using joblib for parallelization didn't go through because we have the futures interface and that one doesn't use joblib.

Reply:
I see, ok. That is fine, agreed. One line config, one line init backend, one line use.

About just joblib: I remember now, thanks.

And to the long-running discussion: I still think that we should ditch futures in favour of workers+queues. There's no good way to interrupt otherwise, or to send information back to samplers (e.g. for adaptive stratified sampling). Both are requirements of several algorithms.

Plus with workers we reuse processes more efficiently and probably transfer less data (or read/copy less often from local storage / ram). And with a decent job/worker object with a couple of hooks built in, algorithm code can remain as simple as it is now. Just subclass and fill in the gaps with your function, and you're good to go.

Reply:
I am not against the idea of using queues, I am just worried that it would bring a new type of bugs that we have to worry about. We should do it properly and take inspiration from other projects. I like for example [Optuna](https://github.com/optuna/optuna/tree/master)'s parallelization approach that uses a database to completely decouple the main process from the workers. They also have samplers that even though are not the same as ours, can still give us an idea of how to implement them.

For now though, I think we should just clean things up for the next release and for pydata.

Reply:
I haven't looked at optuna, but I would assume they use a database for an altogether different reason than decoupling. Namely, keeping track of hyperparameter choices and experiment results. We don't want to keep track of these things, or if we do, they are associated to a result set, which we don't persist. Introducing a big external dependency which requires deployment of a service (and thus might be problematic in many environments) just to avoid keeping track of a couple of queues seems overkill to me. I hope we can find a better, and simpler solution

Reply:
They use the database for both purposes. In the case of queues, we also need to deploy an external dependency.
Anyway, I will prepare a prototype or at least a design document before mentioning that again.